pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -45,12 +47,108 @@ func NewPacker(ec2 ec2iface.EC2API) *PodPacker {
-			Pods:                pods,
-		},
-	}, nil
+
+	// 1. Arrange pods in decreasing order by the amount of CPU requested, if
+	// CPU requested is equal compare memory requested.
+	sort.Sort(sort.Reverse(byResourceRequested{pods}))
+
+	// 2. Get all available instance types with there capacity and cost
+	// TODO add filters
+	instanceTypes := p.getInstanceTypes("")
+	// TODO reserve (Kubelet + daemon sets) overhead for instance types
+	return p.packSortedPods(pods, instanceTypes)
+}
+
+// takes a list of pods sorted based on their resource requirements compared by CPU and memory.
+
+func (p *PodPacker) packSortedPods(pods []*v1.Pod, instanceTypes []*instanceType) ([]*Packings, error) {
+	// Start with the smallest instance type and the biggest pod check how
+	// many pods can we fit, go to the next bigger type and check if we can fit
+	// more pods. Compare pods packed on all and select the instance type with
+	// highest pod count.
+	podsPacked := map[*v1.Pod]bool{}
+	packings := make([]*Packings, 0)
+	for start, pod := range pods {
+		if podsPacked[pod] {
+			continue
+		}
+
+		prevPacking := make([]*v1.Pod, 0)
+		instanceTypesSelected := make([]*instanceType, 0)
+		// Go to the next instance type see how many more pods can we fit?
+		for _, instance := range instanceTypes {
+			// from start index loop through all the pods and check how many
+			// pods we can fit on each instance type
+			packings, err := calculateBestPackingOption(instance, start, pods, podsPacked)
+			if err != nil {
+				zap.S().Errorf("Failed to calculate packing for instance type %s, err, %w", instance.name, err)
+				continue
+			}
+			zap.S().Info("packings are ", len(packings))
+			// If the pods packed are the same as before, this instance type can be
+			// considered a backup option in case we get ICE
+			if len(packings) == len(prevPacking) &&
+				podsMatch(prevPacking, packings) {
+				instanceTypesSelected = append(instanceTypesSelected, instance)
+			} else if len(packings) > len(prevPacking) {
+				// If pods packed are more than compared to what we did last time,
+				// consider using this instance type
+				prevPacking = packings
+				instanceTypesSelected = []*instanceType{instance}
+			}
+		}
+		// checked all instance type and found no packing option
+		if len(prevPacking) == 0 {
+			return nil, fmt.Errorf("no instance type found for packing %d pods", len(pods))
+		}
+		// keep a track of pods we have packed
+		for _, pod := range prevPacking {
+			podsPacked[pod] = true
+		}
+		instanceOptions := []string{}
+		for _, instanceType := range filterInstancesBasedOnCost(instanceTypesSelected) {
+			instanceOptions = append(instanceOptions, instanceType.name)
+		}
+		zap.S().Info("Instance Options are %v", instanceOptions)
+		packings = append(packings, &Packings{prevPacking, instanceOptions})
+	}
+	return packings, nil
+}
+
+func podsMatch(first, second []*v1.Pod) bool {
+	if len(first) != len(second) {
+		return false
+	}
+	podSeen := map[*v1.Pod]struct{}{}
```

Comment:
more idiomatic to use a bool?

Reply:
empty structs are no memory allocation and are idiomatic too https://github.com/golang/go/blob/master/src/net/http/server.go#L3090